### ISSUE TITLE
ubuntu 1604LTS - pip install primesieve hangs on python 3.7.1 fine in 3.7.0 and 3.7.2

### ISSUE THREAD

**Issue by fenchu** [kimwalisch/primesieve#69]
This just hangs when I try to install on python 3.7.1 on ubuntu 1604LTS:
```
pip install primesieve
Collecting primesieve
  Using cached https://files.pythonhosted.org/packages/43/4f/cd9b85d154bc94eec5d4a8cfdbbf96d0e487cecd46495f86b9d5d481d18d/primesieve-1.4.4.tar.gz
Installing collected packages: primesieve
  Running setup.py install for primesieve ... -
```
Please create a wheel of it so I do not have to install 10GB of Visual Studio Bloat to compile on windows.

Thanks

**Comment by kimwalisch** [comment]
Hi,

> Please create a wheel of it so I do not have to install 10GB of Visual Studio Bloat to compile on windows.

Actually, for Windows there are primesieve wheels (compiled using Visual Studio) for Python 2.7, 3.4, 3.5, 3.6 and 3.7 as can be seen here: https://pypi.org/project/primesieve/#files. I just tested installing primesieve using pip and Python 3.7.2 on Windows 10 x64 and it worked fine for me, see the screen blow:

![screenshot](https://user-images.githubusercontent.com/4656219/50451991-c5545380-0937-11e9-8d4b-cd9788310ee8.png)

For Linux and macOS however there are no wheels and hence you need to install a C++ compiler as a prerequisite (if you are using the Windows Subsystem for Linux (WSL) you need to follow the Linux installation instructions). On Linux this should take no more than 10 seconds. (On macOS this can take a very long time which is not ideal...). 

Here is how to install a C++ compiler on Ubuntu:

```bash
# Ubuntu/Debian
$ sudo apt install g++ python-dev
```

Then try reinstalling primesieve using:

```bash
$ pip install primesieve
```

I hope this helps.

**Comment by fenchu** [comment]
Thanks for the wheel info for windows, I will look better in downloads :-)

But when it comes to installing on 3.7.1 it works on 3.7.0 and 3.7.2 but not 3.7.1.

### 3.7.0 is fine
```
$ pyenv virtualenv 3.7.0 primesieve3.7.0
Looking in links: /tmp/tmpanrbdbav
Requirement already satisfied: setuptools in /home/mortenb/.pyenv/versions/3.7.0/envs/primesieve3.7.0/lib/python3.7/site-packages (39.0.1)
Requirement already satisfied: pip in /home/mortenb/.pyenv/versions/3.7.0/envs/primesieve3.7.0/lib/python3.7/site-packages (10.0.1)
$ pyenv activate primesieve3.7.0 
pyenv-virtualenv: prompt changing will be removed from future release. configure `export PYENV_VIRTUALENV_DISABLE_PROMPT=1' to simulate the behavior.
(primesieve3.7.0) mortenb@mortenlap4:/dist/python-dev/primesieve$ pip install primesieve
Collecting primesieve
  Using cached https://files.pythonhosted.org/packages/43/4f/cd9b85d154bc94eec5d4a8cfdbbf96d0e487cecd46495f86b9d5d481d18d/primesieve-1.4.4.tar.gz
Installing collected packages: primesieve
  Running setup.py install for primesieve ... done
Successfully installed primesieve-1.4.4
You are using pip version 10.0.1, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
(primesieve3.7.0) mortenb@mortenlap4:/dist/python-dev/primesieve$ python
Python 3.7.0 (default, Dec 27 2018, 11:50:57) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import primesieve
>>> primesieve.primes(100)
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]
>>> exit()
```

### 3.7.1 hangs
```
$ pyenv virtualenv 3.7.1 primesieve3.7.1
Looking in links: /tmp/tmp975x6gnr
Requirement already satisfied: setuptools in /home/mortenb/.pyenv/versions/3.7.1/envs/primesieve3.7.1/lib/python3.7/site-packages (39.0.1)
Requirement already satisfied: pip in /home/mortenb/.pyenv/versions/3.7.1/envs/primesieve3.7.1/lib/python3.7/site-packages (10.0.1)
$ pyenv activate primesieve3.7.1 
pyenv-virtualenv: prompt changing will be removed from future release. configure `export PYENV_VIRTUALENV_DISABLE_PROMPT=1' to simulate the behavior.
$ pip install primesieve
Collecting primesieve
  Using cached https://files.pythonhosted.org/packages/43/4f/cd9b85d154bc94eec5d4a8cfdbbf96d0e487cecd46495f86b9d5d481d18d/primesieve-1.4.4.tar.gz
Installing collected packages: primesieve
  Running setup.py install for primesieve ... -
<HANGS HERE>
^canceled
Operation cancelled by user
```

### and 3.7.2 is fine
```
$ pyenv virtualenv 3.7.2 primesieve3.7.2
Looking in links: /tmp/tmpf294cwcs
Requirement already satisfied: setuptools in /home/mortenb/.pyenv/versions/3.7.2/envs/primesieve3.7.2/lib/python3.7/site-packages (40.6.2)
Requirement already satisfied: pip in /home/mortenb/.pyenv/versions/3.7.2/envs/primesieve3.7.2/lib/python3.7/site-packages (18.1)
(primesieve3.7.1) mortenb@mortenlap4:/dist/python-dev/primesieve$ pyenv activate primesieve3.7.2
pyenv-virtualenv: prompt changing will be removed from future release. configure `export PYENV_VIRTUALENV_DISABLE_PROMPT=1' to simulate the behavior.
(primesieve3.7.2) mortenb@mortenlap4:/dist/python-dev/primesieve$ pip install primesieve
Collecting primesieve
  Using cached https://files.pythonhosted.org/packages/43/4f/cd9b85d154bc94eec5d4a8cfdbbf96d0e487cecd46495f86b9d5d481d18d/primesieve-1.4.4.tar.gz
Installing collected packages: primesieve
  Running setup.py install for primesieve ... done
Successfully installed primesieve-1.4.4
$ python
Python 3.7.2 (default, Dec 27 2018, 11:30:31) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import primesieve
>>> primesieve.primes(100)
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]
>>> exit()
```

**Comment by kimwalisch** [comment]
Can you please try to run the commands below using Python 3.7.1 and report your terminal output?

```bash
# Here is how to debug the Python 3.7.1 issue

wget https://files.pythonhosted.org/packages/43/4f/cd9b85d154bc94eec5d4a8cfdbbf96d0e487cecd46495f86b9d5d481d18d/primesieve-1.4.4.tar.gz
tar xvf primesieve-1.4.4.tar.gz
cd primesieve-1.4.4
pip install . --upgrade --force-reinstall --verbose
```

Note that the official repo for the primesieve python bindings is https://github.com/hickford/primesieve-python. But as I cannot transfer this issue to that repo I will try to resolve the issue here.

**Comment by kimwalisch** [comment]
I have now been able to reproduce your issue on Ubuntu 16.04 with Python 3.7.1 compiled from source:

```bash
ST.in'
    writing manifest file 'pip-egg-info/primesieve.egg-info/SOURCES.txt'
    /usr/local/lib/python3.7/distutils/dist.py:274: UserWarning: Unknown distribution option: 'long_description_content_type'
      warnings.warn(msg)
  Source in /tmp/pip-install-f_m2gbsf/primesieve has version 1.4.4, which satisfies requirement primesieve from https://files.pythonhosted.org/packages/43/4f/cd9b85d154bc94eec5d4a8cfdbbf96d0e487cecd46495f86b9d5d481d18d/primesieve-1.4.4.tar.gz#sha256=58bbff5c6928254e4fadd21ce324d8bae97aa706f72d69b1b0e1bf7ed6a1925c
Installing collected packages: primesieve
  Created temporary directory: /tmp/pip-record-x22efaha
  Running setup.py install for primesieve ...     Running command /usr/local/bin/python3.7 -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-f_m2gbsf/primesieve/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-x22efaha/install-record.txt --single-version-externally-managed --compile
    get_compiler_openmp_flag(): -fopenmp
    /usr/local/lib/python3.7/distutils/dist.py:274: UserWarning: Unknown distribution option: 'long_description_content_type'
      warnings.warn(msg)
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.7
    creating build/lib.linux-x86_64-3.7/primesieve
    copying primesieve/__init__.py -> build/lib.linux-x86_64-3.7/primesieve
    creating build/lib.linux-x86_64-3.7/primesieve/numpy
    copying primesieve/numpy/__init__.py -> build/lib.linux-x86_64-3.7/primesieve/numpy
    running build_ext
    building 'primesieve._primesieve' extension
    creating build/temp.linux-x86_64-3.7
    creating build/temp.linux-x86_64-3.7/primesieve
    creating build/temp.linux-x86_64-3.7/lib
    creating build/temp.linux-x86_64-3.7/lib/primesieve
    creating build/temp.linux-x86_64-3.7/lib/primesieve/src
    creating build/temp.linux-x86_64-3.7/lib/primesieve/src/primesieve

# hangs
```

The bug is caused by:

```bash
/usr/local/lib/python3.7/distutils/dist.py:274: UserWarning:
Unknown distribution option: 'long_description_content_type'
```

The relevant section in the ```setup.py``` file is:

```Python
setup(
    # ...
    long_description = open('README.md',"rb").read().decode('utf8'),
    long_description_content_type = 'text/markdown',
    # ...
)
```

**Comment by kimwalisch** [comment]
Ideally we should check inside ```setup.py``` whether the user's python supports ```long_description``` before using it.

**Comment by kimwalisch** [comment]
When I cancel the primesieve installation using Ctrl+C on macOS I get the following Traceback:

```bash
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/cli/base_command.py", line 154, in main
    status = self.run(options, args)
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/commands/install.py", line 336, in run
    session=session, autobuilding=True
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/wheel.py", line 787, in build
    python_tag=python_tag,
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/wheel.py", line 646, in _build_one
    python_tag=python_tag)
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/wheel.py", line 650, in _build_one_inside_env
    if self.__build_one(req, temp_dir.path, python_tag=python_tag):
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/wheel.py", line 689, in __build_one
    show_stdout=False, spinner=spinner)
  File "/usr/local/lib/python3.7/site-packages/pip/_internal/utils/misc.py", line 670, in call_subprocess
    line = console_to_str(proc.stdout.readline())
KeyboardInterrupt
/usr/local/lib/python3.7/subprocess.py:852: ResourceWarning: subprocess 77421 is still running
  ResourceWarning, source=self)
```

The installation seems to be waiting for a subprocess to finish. Unfortunately I don't have enough Python knowledge to debug this issue.

This bug does not seem related to primesieve so I think there is a bug in Python 3.7.1 which causes the installation to hang. Also that Python bug seems to have been fixed in Python 3.7.2 since there are no primesieve installation issues using Python 3.7.2.

I will close this issue for now as there is nothing I can do.
